convert/convert.go

```diff
@@ -0,0 +1,140 @@
+	opType, ok := oplogEntry["op"].(string)
+	if !ok {
+		return nil, fmt.Errorf("Missing op type")
+	}
+	namespace, ok := oplogEntry["ns"].(string)
+	if !ok {
+		return nil, fmt.Errorf("Missing namespace")
+	}
+
+	// Ignore changes to the system namespace. These are things like system.indexes
+	if strings.HasPrefix(namespace, "system.") {
+		return nil, nil
+	}
+
+	obj, ok := oplogEntry["o"].(bson.M)
+	if !ok {
+		return nil, fmt.Errorf("Missing object field")
+	}
+
+	switch opType {
+	case "i":
+		return convertToInsert(namespace, obj)
+	case "u":
+		return convertToUpdate(namespace, obj, oplogEntry)
+	case "d":
+		return convertToRemove(namespace, obj, oplogEntry)
+	default:
+		// It's theoretically possibly that is also 'c' or 'n', but we don't support them so let's error out
+		return nil, fmt.Errorf("Unknown op type %s", opType)
+	}
+}
+
+func convertToInsert(namespace string, obj bson.M) (*operation.Op, error) {
+	op := operation.Op{Namespace: namespace, Type: "insert"}
+	id, ok := obj["_id"]
+	if !ok {
+		return nil, fmt.Errorf("Insert missing or 'o._id' field")
+	}
+
+	var err error
+	op.ID, err = convertIdToString(id)
+	if err != nil {
+		return nil, err
+	}
+	op.Obj = obj
+	return &op, nil
+}
+
+func convertToUpdate(namespace string, obj, oplogEntry bson.M) (*operation.Op, error) {
+	op := operation.Op{Namespace: namespace, Type: "update"}
+	id, ok := oplogEntry["o2"].(bson.M)["_id"]
+	if !ok {
+		return nil, fmt.Errorf("Update missing o._id field")
+	}
+
+	var err error
+	op.ID, err = convertIdToString(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check to make sure the object only has $ fields we understand
+	// Note that other Mongo update commands (afaict) are converted to either direct
+	// set commands or $set and $unset commnands. For example an $addToSet command
+	// becoomes {"$set" : {"key.1" : "value"}}
+	for key := range obj {
+		if strings.Contains(key, "$") && key != "$set" && key != "$unset" {
+			return nil, fmt.Errorf("Invalid key %s in update object", key)
+		}
+	}
+	op.Obj = obj
+
+	// Since this field is referenced in the Mongo applyCmd source code, but I haven't been able to
+	// set it in any of our oplog entries, let's just sanity check that it isn't set.
+	if _, ok = oplogEntry["b"]; ok {
+		return nil, fmt.Errorf("Unknown field 'b' in update")
+	}
+	return &op, nil
+}
+
+func convertToRemove(namespace string, obj, oplogEntry bson.M) (*operation.Op, error) {
+	op := operation.Op{Namespace: namespace, Type: "remove"}
+	id, ok := obj["_id"]
+	if !ok {
+		return nil, fmt.Errorf("Delete missing '_id' field")
+	}
+
+	var err error
+	op.ID, err = convertIdToString(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// We see this on all our deletes so let's keep making sure it's there
+	if b, ok := oplogEntry["b"].(bool); !ok || !b {
+		return nil, fmt.Errorf("'b' field not set to true for delete")
+	}
+	return &op, nil
+}
+
+func convertIdToString(id interface{}) (string, error) {
+	if str, ok := id.(string); ok {
```

Comment:
Mongo ids don't have to be object ids (http://stackoverflow.com/questions/1352671/unique-ids-with-mongodb). I handled object ids and strings because that's what I've seen in our oplogs. I could try to support more, but I'm hesitant to support more until we need them since every one adds the chance of doing something wrong. Let me know if you disagree. I will definitely make this clear in a comment.